api/br/v1alpha1/tibr_types.go

```diff
@@ -0,0 +1,90 @@
+// +k8s:openapi-gen=true
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+// +kubebuilder:resource:categories=br
+// +kubebuilder:resource:shortName="tibr"
+// +kubebuilder:selectablefield:JSONPath=`.spec.cluster.name`
+// +kubebuilder:printcolumn:name="Cluster",type=string,JSONPath=`.spec.cluster.name`
+// +kubebuilder:printcolumn:name="ScheduleType",type=string,JSONPath=`.spec.autoSchedule.type`
+// +kubebuilder:printcolumn:name="ScheduleAt",type=string,JSONPath=`.spec.autoSchedule.at`
+// +kubebuilder:printcolumn:name="Synced",type=string,JSONPath=`.status.conditions[?(@.type=="Synced")].status`
+// +kubebuilder:printcolumn:name="Ready",type=string,JSONPath=`.status.conditions[?(@.type=="Ready")].status`
+// +kubebuilder:printcolumn:name="Age",type="date",JSONPath=".metadata.creationTimestamp"
+
+// TiBR is the Schema for the tibr API
+type TiBR struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   TiBRSpec   `json:"spec,omitempty"`
+	Status TiBRStatus `json:"status,omitempty"`
+}
+
+// TiBRSpec defines the desired state of TiBR.
+type TiBRSpec struct {
+	// Cluster is a reference of tidb cluster
+	Cluster v1alpha1.ClusterReference `json:"cluster"`
+	// AutoSchedule define auto backup strategy for TiBR, it's optional
+	AutoSchedule *TiBRAutoSchedule `json:"autoSchedule,omitempty"`
+	// Image is image of br service, default is pingcap/tikv
+	Image  *string             `json:"image,omitempty"`
+	Config v1alpha1.ConfigFile `json:"config,omitempty"`
+	// Overlay defines a k8s native resource template patch
+	Overlay *v1alpha1.Overlay `json:"overlay,omitempty"`
+}
+
+// TiBRStatus defines the observed state of TiBR.
+type TiBRStatus struct {
+	v1alpha1.CommonStatus `json:",inline"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:object:root=true
+
+// TiBRList contains a list of TiBR.
+type TiBRList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []TiBR `json:"items"`
+}
+
+type TiBRAutoScheduleType string
+
+const (
+	TiBRAutoScheduleTypePerDay    TiBRAutoScheduleType = "per-day"
+	TiBRAutoScheduleTypePerHour   TiBRAutoScheduleType = "per-hour"
+	TiBRAutoScheduleTypePerMinute TiBRAutoScheduleType = "per-minute"
+)
+
+// +kubebuilder:validation:XValidation:rule="(self.type== 'per-day' && self.at >= 0 && self.at <= 23)|| (self.type == 'per-hour' && self.at >= 0 && self.at <= 59) || (self.type == 'per-minute' && self.at == 0)"
+// +kubebuilder:validation:Message="if type is per-day, at is the hour of day; if type is per-hour, at is the minute of hour; if type is per-minute, it should be 0"
+type TiBRAutoSchedule struct {
+	// Type defines the schedule type, such as per-day, per-hour, per-minute
+	// +kubebuilder:validation:Enum=per-day;per-hour;per-minute
+	Type TiBRAutoScheduleType `json:"type"`
+	// At defines the schedule time of backup
+	At uint32 `json:"at"`
```

Comment:
Give an example?

Reply:
I think this explain very detail in kubebuilder annotation 
// +kubebuilder:validation:XValidation:rule="(self.type== 'per-day' && self.at >= 0 && self.at <= 23)|| (self.type == 'per-hour' && self.at >= 0 && self.at <= 59) || (self.type == 'per-minute' && self.at == 0)"
// +kubebuilder:validation:Message="if type is per-day, at is the hour of day; if type is per-hour, at is the minute of hour; if type is per-minute, it should be 0"